ssz/sedes/base.py

```diff
@@ -104,77 +83,126 @@ def __init__(self, size: int):
+    return sum(
+        sedes.get_fixed_size() if sedes.is_fixed_sized else constants.OFFSET_SIZE
+        for sedes in element_sedes
+    )
+
 
+class CompositeSedes(BaseCompositeSedes[TSerializable, TCanonical]):
     @abstractmethod
-    def deserialize_content(self, content: bytes) -> TDeserialized:
+    def _iter_sedes(self,
+                    value: Sequence[TSerializable],
+                    ) -> Tuple[Tuple[TSerializable, TSedes], ...]:
         pass
 
-    #
-    # Tree hashing
-    #
-    def hash_tree_root(self, value: TSerializable) -> bytes:
-        serialized_value = self.serialize(value)
-        return merkleize(pack((serialized_value,)))
+    def _validate_serializable(self, value: Any) -> None:
+        pass
 
+    def serialize(self, value: Sequence[TSerializable]) -> bytes:
+        self._validate_serializable(value)
 
-class CompositeSedes(BaseSedes[TSerializable, TDeserialized]):
+        if not len(value):
+            return b''
 
-    #
-    # Serialization
-    #
-    def serialize(self, value: TSerializable) -> bytes:
-        content = self.serialize_content(value)
+        pairs = self._iter_sedes(value)
+        element_sedes = tuple(sedes for element, sedes in pairs)
 
-        if self.is_static_sized:
-            return content
+        section_0_size = _compute_section_0_size(element_sedes)
+
+        section_1_components = tuple(
+            sedes.serialize(item)
+            for item, sedes
+            in pairs
+            if not sedes.is_fixed_sized
+        )
+
+        if section_1_components:
+            offsets = tuple(accumulate(
+                operator.add,
+                map(len, section_1_components[:-1]),
+                section_0_size,
+            ))
         else:
-            validate_content_size(content)
-            size_prefix = get_size_prefix(content)
-            return size_prefix + content
+            offsets = ()
+
+        offsets_iter = iter(offsets)
+
+        section_0_components = tuple(
+            sedes.serialize(item) if sedes.is_fixed_sized else encode_offset(next(offsets_iter))
+            for item, sedes
+            in pairs
+        )
+
+        try:
+            next(offsets_iter)
+        except StopIteration:
+            pass
+        else:
+            raise DeserializationError("Did not consume all offsets while decoding value")
+
+        return b"".join(concatv(
+            section_0_components,
+            section_1_components,
+        ))
+
+    def deserialize(self, data: bytes) -> TCanonical:
+        stream = io.BytesIO(data)
+        value = self.s_deserialize(stream)
+        extra_data = stream.read()
+        if extra_data:
+            raise DeserializationError(f"Got {len(extra_data)} superfluous bytes")
+        return value
 
     @abstractmethod
-    def serialize_content(self, value: TSerializable) -> bytes:
+    def s_deserialize(self, stream: IO[bytes]) -> TCanonical:
```

Comment:
The change to using streaming here is intended to just be a 1:1 replacement of the previous fragile approach where the various serialization methods returned the new position in the data where subsequent fields should read from.  I posit that these two approaches are equivalent with the stream based method being less error prone since it doesn't require manual bookkeeping to track the current position in the data.

Here is an example of the historical code:

https://github.com/ethereum/py-ssz/blob/f9afed6439046e417bbdc399eaf3ade5481b7d6c/ssz/sedes/base.py#L130-L132

Using streams, this becomes:

```python
def deserialize_segment(self, stream: IO[bytes]) -> TDeserialized:
        content = stream.read(self.size)
        return self.deserialize_content(content)
```

The new stream based decoding API is intended to be private but I realize I didn't name it as such so I'll update the name to be `_deserialize_stream` or something similar.

Reply:
Got it, now I agree that using streams is better than passing around indices.